packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,66 +168,407 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+
+Kotlin:       1.7.10
+Groovy:       3.0.13
+Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
+JVM:          17.0.6 (Homebrew 17.0.6+0)
+OS:           Mac OS X 13.2.1 aarch64
+    */
+    // Observation shows that the version can have 2 or 3 numbers.
+    // Inner parentheticals `(\.\d+)?` denote the optional third value.
+    // Outter parentheticals `Gradle (...)` denote a grouping used to extract
+    // the version number.
+    final RegExp gradleVersionRegex = RegExp(r'Gradle (\d+\.\d+(\.\d+)?)');
+    final RegExpMatch? version =
+        gradleVersionRegex.firstMatch(gradleVersionVerbose);
+    if (version == null) {
+      // Most likley a bug in our parse implementation/regex.
+      logger.printWarning(_formatParseWarning(gradleVersionVerbose));
+      return null;
+    }
+    return version.group(1);
+  } else {
+    logger.printTrace('Could not run system gradle');
+    return null;
+  }
 }
 
+/// Returns the Android Gradle Plugin (AGP) version that the current project
+/// depends on when found, null otherwise.
+///
+/// The Android plugin version is specified in the [build.gradle] file within
+/// the project's Android directory ([androidDirectory]).
+String? getAgpVersion(Directory androidDirectory, Logger logger) {
+  final File buildFile = androidDirectory.childFile('build.gradle');
+  if (!buildFile.existsSync()) {
+    logger.printTrace('Can not find build.gradle in $androidDirectory');
+    return null;
+  }
+  final String buildFileContent = buildFile.readAsStringSync();
+  final Iterable<Match> pluginMatches =
+      _androidGradlePluginRegExp.allMatches(buildFileContent);
+  if (pluginMatches.isEmpty) {
+    logger.printTrace("$buildFile doesn't provide an AGP version");
+    return null;
+  }
+  final String? androidPluginVersion = pluginMatches.first.group(1);
+  logger.printTrace('$buildFile provides AGP version: $androidPluginVersion');
+  return androidPluginVersion;
+}
+
+String _formatParseWarning(String content) {
+  return 'Could not parse gradle version from: \n'
+      '$content \n'
+      'If there is a version please look for an existing bug '
+      'https://github.com/flutter/flutter/issues/'
+      ' and if one does not exist file a new issue.';
+}
+
+// Validate that Gradle version and AGP are compatible with each other.
+//
+// Returns true if versions are compatible.
+// Null Gradle version or AGP version returns false.
+// If compatability can not be evaulated returns false.
+// If versions are newer than the max known version a warning is logged and true
+// returned.
+//
+// Source of truth found here:
+// https://developer.android.com/studio/releases/gradle-plugin#updating-gradle
+// AGP has a minimim version of gradle required but no max starting at
+// AGP version 2.3.0+.
+bool validateGradleAndAgp(Logger logger,
+    {required String? gradleV, required String? agpV}) {
+
+  // TODO(reidbaker): are these the correct values we want to support?
```

Comment:
Is this a TODO to land, or to resolve during review? If the former, it's not really clear what the criteria for someone removing/fixing the TODO would be.

Reply:
This LGTM. If we ever run this during a `flutter build`, we should just make sure there's an opt out flag.